File: Interviews/Amazon/GrayCheck.java
/*
 * Amazon OA: Gray Check
 * Check if two bytes could be placed consectively in a gray code sequence
 *
 **/

public class GrayCheck{
    public static int grayCheck(byte b1, byte b2){
        byte ret = (byte)(b1 ^ b2);
        int cnt = 0;

        while(ret > 0){
            cnt ++;
            ret &= (ret - 1);
        }

        return (cnt == 1) ? 1 : -1;
    }

    public static void test(byte b1, byte b2){
        if(grayCheck(b1, b2) == 1)
            System.out.println("Yes they could");
        else
            System.out.println("No they could not");
    }
    
    
    public static void main(String[] args){
        byte b1 = 0x01;
        byte b2 = 0x02;

        test(b1, b2);  // no

        b1 = 0x03;
        test(b1, b2);  // yes
        
        b1 = 0x00;
        test(b1, b2); // yes
    }
};

File: Interviews/Amazon/gray_check.cc
/*
 * =====================================================================================
 *
 *       Filename:  gray_check.cc
 *
 *    Description:  check if two bytes could be placed successively in a sequence of 
 *                  gray codes
 *                  
 *                  (That is these two bytes only differ in 1 bit)
 *
 *        Created:  03/01/2015 15:41:45
 *       Compiler:  g++ 4.7.0
 *
 * =====================================================================================
 */

#include <iostream>
#include <cassert>

using namespace std;

int grayCheck(unsigned char term1, unsigned char term2){
    int term = term1 ^ term2;
    int num = 0;

    while(term > 0){
        num++;
        term = term & (term - 1);
    }

    return (num == 1) ? 1 : 0;
}

void test(unsigned char term1, unsigned char term2){
    if(grayCheck(term1, term2) == 1){
        cout << "Could be placed!" << endl;
    }
    else{
        cout << "Couldn't be placed!" << endl;
    }
}

int main(){
    unsigned char term1 = 0b0111111;
    unsigned char term2 = 0b0111110;

    test(term1, term2);  // could

    term1 &= 0;
    test(term1, term2);  // couldn't
    
    term1 = 0b00000001;
    term2 = 0b00000010;
    test(term1, term2);  // couldn't
    
    term1 &= 0;
    test(term1, term2);  // could

    return 0;
}

